### ISSUE TITLE
(optionally) use tar instead of zip to send directories

### ISSUE THREAD

**Issue by danboid** [warner/magic-wormhole#396]
It seems wormhole zips directories but pkzip files don't retain Linux file permissions like tar files do so it would be good to see tar support added.

**Comment by noproto** [comment]
There are many security related considerations in the tar archive format. I'm not a maintainer but I'd like to point that out for discussion.

**Comment by piegamesde** [comment]
For <https://github.com/magic-wormhole/magic-wormhole-protocols/pull/1>, I did some research on different archive formats and settled on tar as well. What are the security considerations (and links perhaps?) and do you know of any better format?

**Comment by noproto** [comment]
Tar archives natively support being able to create symbolic links and block devices. Unix ZIP archives technically support that too, but the extract() function of Python's zipfile library (currently used by magic-wormhole) won't unpack these symlinks. If tar is used (along with offer type "directory"), you can get symlinks in the created destination directory. I haven't verified this yet but in theory it might be able to bypass the path filters and traverse into arbitrary directories, the way its currently designed (I'm most likely wrong but it would need to be tested as part of any implementation). Regarding other special files, the tar format has the ability to create block devices such as those that are in /dev/ (think making a file that allows your entire disk to be writable), and possibly more edge cases like SETUID/SETGID binaries that would be useful to a local attacker seeking elevated privileges.

**Comment by meejah** [comment]
My personal opinion is that it would be better overall to augment the file-transfer protocol so that multiple files can be transmitted (possibly in both directions) once the connection is established (since "establish the connection" is the "hard part" as far as the humans are concerned).

This of course may come with its own security considerations, but "magically zip/tar/whatever" files up seems like at least partially a hack around the fact that there's only one "thing" transferred. UX-wise of course it still makes sense that people might like to "transfer this directory"...

**Comment by piegamesde** [comment]
There are pretty good arguments for sending multiple files or whole directories in one transfer, which is best done through some file archive format. The need to having multiple transfers is a separate one.

**Comment by truatpasteurdotfr** [comment]
Why not leaving the choice of zip/tar/... to the end user if multiple files are to be transfered? One can also debate which compression is "best", among zip/gzip/pigz/... :P

**Comment by piegamesde** [comment]
What for? If users care about the format, they can simply send their file archive as such right away. The discussion is around what happens when a user sends a *directory*. A directory is sent, a directory is received, and the transport container is an implementation detail. Users won't (and shouldn't) care about this (if it's done well), so why should we ask them?

Regarding the compression format, that's a different story.

**Comment by meejah** [comment]
> There are pretty good arguments for sending multiple files or whole directories in one transfer, which is best done through some file archive format. 

I agree that sending multiple files / "a directory" is a good use-case.

I guess we seem to disagree slightly on the method? By "one transfer" are you discounting things like e.g. a "v2" protocol that allows multiple files (which might technically be >1 network connection, but will still be just one "wormhole code" to set up)...?

**Comment by piegamesde** [comment]
For me, a "transfer" from the user's perspective is a proposition to send some files together with an ack from the other side. If the files are acked as one, it is one transfer. If there are multiple transfer offers that are accepted/rejected individually, it's multiple transfers.

Terminology gets muddy when dealing with connection losses / resuming transfers. In that case, it kind of depends on how we'll implement it.

**Comment by meejah** [comment]
Okay .. maybe it makes sense to refer to "a transfer session" as the thing which is set up with a single human interaction i.e. one exchange of wormhole codes.

Considering "a transfer" to be one offer/ack pair sounds reasonable to me.

Thus, a singe Transfer Session may have one or more Transfers occur within it. The current command-line does "a single Transfer", for example (even though it might actually be more than a single file).

A use-case that includes multiple Transfers in a single Transfer Session might be a GUI case: set up a session by exchanging a code and then drag/drop files -- each such "drop" action might result in a Transfer (where the UX could choose to confirm "receive X files?" each time). Participants might leave the windows open for considerable time -- e.g. they are having a phone call and keep remembering more data to send each other.

**Comment by al12gamer** [comment]
...so basically uploading a .tar.xz file doesn't work because of this?

**Comment by meejah** [comment]
@al12gamer you can absolutely transfer a `.tar.xz` file (or anything else).

This discussion is about what happens when you indicate on the CLI that you wish to transfer a directory (e.g. "`~/public_html/`" or something). In that case, with the existing transfer protocol, it packages the files up and sends them as "an archive".

There's also a next-generation transfer protocol, spec and implementations happening which will change the above at the technical level. The user would still "transfer the directory `~/public_html`" (for example) but a mechanism different from "literally make a `.zip` file" would likely be used to accomplish that.

**Comment by Skylion007** [comment]
I would recommend tar.zstd files as it's uniquely well suited to cases where the file is compressed / decompressed once (since it's really fast while also being a performant compression algorithm).
